test/tesla/adapter/gun_test.exs

```diff
@@ -41,10 +42,114 @@ defmodule Tesla.Adapter.GunTest do
     }
 
     assert {:ok, %Env{} = response} = call(request)
     assert response.status == 200
+    assert byte_size(response.body) == 16
+  end
+
+  test "read response body in chunks" do
+    request = %Env{
+      method: :get,
+      url: "#{@http}/stream-bytes/10"
+    }
+
+    assert {:ok, %Env{} = response} = call(request, body_as: :chunks)
+    assert response.status == 200
+    %{pid: pid, stream: stream, opts: opts} = response.body
+    assert opts[:body_as] == :chunks
+    assert is_pid(pid)
+    assert is_reference(stream)
+
+    assert read_body(pid, stream) |> byte_size() == 16
+    refute Process.alive?(pid)
+  end
+
+  test "read response body in stream" do
+    request = %Env{
+      method: :get,
+      url: "#{@http}/stream-bytes/10"
+    }
+
+    assert {:ok, %Env{} = response} = call(request, body_as: :stream)
+    assert response.status == 200
+    assert is_function(response.body)
+    assert Enum.to_list(response.body) |> List.to_string() |> byte_size() == 16
+  end
+
+  test "read response body in stream with opened connection without closing connection" do
+    uri = URI.parse(@http)
+    {:ok, conn} = :gun.open(to_charlist(uri.host), uri.port)
+
+    request = %Env{
+      method: :get,
+      url: "#{@http}/stream-bytes/10"
+    }
+
+    assert {:ok, %Env{} = response} =
+             call(request, body_as: :stream, conn: conn, close_conn: false)
+
+    assert response.status == 200
+    assert is_function(response.body)
+    assert Enum.to_list(response.body) |> List.to_string() |> byte_size() == 16
+
+    assert Process.alive?(conn)
+
+    :ok = Gun.close(conn)
+    refute Process.alive?(conn)
+  end
+
+  test "read response body in stream with opened connection with closing connection" do
+    uri = URI.parse(@http)
+    {:ok, conn} = :gun.open(to_charlist(uri.host), uri.port)
+
+    request = %Env{
+      method: :get,
+      url: "#{@http}/stream-bytes/10"
+    }
+
+    assert {:ok, %Env{} = response} = call(request, body_as: :stream, conn: conn)
+
+    assert response.status == 200
+    assert is_function(response.body)
+    assert Enum.to_list(response.body) |> List.to_string() |> byte_size() == 16
+
+    refute Process.alive?(conn)
+  end
+
+  test "error response" do
+    request = %Env{
+      method: :get,
+      url: "#{@http}/status/500"
+    }
+
+    assert {:ok, %Env{} = response} = call(request, timeout: 1_000)
+    assert response.status == 500
+  end
+
+  test "query without path" do
+    request = %Env{
+      method: :get,
+      url: "#{@http}",
+      query: [
+        param: "value"
+      ]
+    }
+
+    assert {:ok, %Env{} = response} = call(request, timeout: 1_000)
+    assert response.status == 400
+  end
+
+  defp read_body(pid, stream, acc \\ "") do
+    case Tesla.Adapter.Gun.read_chunk(pid, stream, timeout: 1_000) do
+      {:fin, body} ->
+        :ok = Gun.close(pid)
+        acc <> body
+
+      {:nofin, part} ->
+        read_body(pid, stream, acc <> part)
+    end
```

Comment:
It would be nice to add "reuse connection" test

Reply:
[Reused connection test with plain body](https://github.com/teamon/tesla/pull/310/files#diff-395315891a5107d753ddcfe9ab7998e5R83)
[Reading body in chunks with reusing connection](https://github.com/teamon/tesla/pull/310/files#diff-395315891a5107d753ddcfe9ab7998e5R109)
[Reading body as stream with reusing connection](https://github.com/teamon/tesla/pull/310/files#diff-395315891a5107d753ddcfe9ab7998e5R186)